File: src/lib/worker/wiki/tables.ts
/**
 * Tables-wiki agent.
 *
 * Reads the catalog's tables/views/skills and writes interlinked concept
 * pages into wiki_pages with kind='tables'. Two layers of pages:
 *
 *   1. Per-table pages: one wiki page per source table OR view, slug
 *      mirroring the doc path (e.g. tables/public.deviations).
 *   2. Index page (slug='_index'): an overview tying everything together,
 *      grouping tables by their FK-connected clusters.
 *
 * The agent is hash-gated — re-running is cheap when nothing changed.
 * Cross-references use the [[tables/<slug>]] syntax which the wiki link
 * parser picks up.
 *
 * Where the doc body comes from: we reuse the per-table markdown that Loop 2
 * already wrote into `docs.markdown`. The wiki agent's job is to add the
 * "## See also" / "## Used in views" / "## Used by skills" sections that
 * stitch the corpus together.
 */
import { catalogPool } from "../../catalog/db";
import { upsertWikiPage } from "../../catalog/wiki";

type TableInfo = {
  table_id: number;
  schema_name: string;
  table_name: string;
  status: string;
  row_count: number | null;
  is_view: boolean;
  doc_md: string | null;
  fk_targets: string[]; // qualified "schema.table" strings
};

type ViewInfo = { name: string; description: string | null; sql: string };
type SkillInfo = { slug: string; name: string; triggers: string[]; description: string };

export async function runTablesWikiAgent(): Promise<{ generated: number }> {
  const tables = await loadTables();
  const views = await loadViews();
  const skills = await loadSkills();

  // Pre-compute slugs so we can cross-link symmetrically.
  const slugFor = (t: { schema_name: string; table_name: string }) => tableSlug(t.schema_name, t.table_name);
  const tableSlugSet = new Set(tables.map((t) => slugFor(t)));

  let generated = 0;
  const ACTOR = "wiki-agent:tables";

  // --- Per-table / per-view pages ---------------------------------------
  for (const t of tables) {
    const slug = slugFor(t);
    const title = t.is_view
      ? `${t.schema_name}.${t.table_name}  (view)`
      : `${t.schema_name}.${t.table_name}`;

    const body = renderTablePage({ table: t, tableSlugSet, views, skills });
    const r = await upsertWikiPage(ACTOR, {
      kind: "tables",
      slug,
      title,
      summary: t.is_view
        ? `Saved view · ${t.row_count ?? 0} rows`
        : `Table · ${t.row_count ?? 0} rows`,
      body_md: body,
      source_ref: {
        table_id: t.table_id,
        schema: t.schema_name,
        name: t.table_name,
        is_view: t.is_view,
      },
    });
    if (r.action !== "skipped") generated++;
  }

  // --- Index page --------------------------------------------------------
  const indexBody = renderIndexPage(tables, views, skills);
  const ix = await upsertWikiPage(ACTOR, {
    kind: "tables",
    slug: "_index",
    title: "Tables overview",
    summary: `${tables.filter((t) => !t.is_view).length} tables · ${tables.filter((t) => t.is_view).length} saved views`,
    body_md: indexBody,
    source_ref: { tables: tables.length, views: views.length, skills: skills.length },
  });
  if (ix.action !== "skipped") generated++;

  return { generated };
}

// ─── Render helpers ────────────────────────────────────────────────────────

function renderTablePage(args: {
  table: TableInfo;
  tableSlugSet: Set<string>;
  views: ViewInfo[];
  skills: SkillInfo[];
}): string {
  const { table, tableSlugSet, views, skills } = args;

  const parts: string[] = [];
  parts.push(`# ${table.schema_name}.${table.table_name}`);
  parts.push("");

  if (table.doc_md) {
    // Strip the leading title from the existing doc to avoid duplicate H1.
    const stripped = table.doc_md.replace(/^#\s+[^\n]+\n+/, "");
    parts.push(stripped);
    parts.push("");
  } else {
    parts.push(`*The catalog hasn't produced a structural+semantic doc yet — re-tick the scheduler.*`);
    parts.push("");
  }

  // ── See also (FK neighbours) ─────────────────────────────────────────
  const fkNeighbours = table.fk_targets
    .map((f) => {
      // FK strings come in as "schema.table.column"; trim the column.
      const m = f.match(/^([^.]+)\.([^.]+)/);
      if (!m) return null;
      const slug = tableSlug(m[1]!, m[2]!);
      return tableSlugSet.has(slug) ? slug : null;
    })
    .filter((s): s is string => !!s);

  if (fkNeighbours.length > 0) {
    parts.push("## See also");
    parts.push("");
    for (const s of fkNeighbours) parts.push(`- [[tables/${s}]]`);
    parts.push("");
  }

  // ── Used in views ────────────────────────────────────────────────────
  const usedInViews = views.filter((v) =>
    new RegExp(`\\b${escapeReg(table.table_name)}\\b`, "i").test(v.sql),
  );
  if (usedInViews.length > 0) {
    parts.push("## Used in views");
    parts.push("");
    for (const v of usedInViews) {
      const vslug = tableSlug("loom_views", v.name);
      const linkable = tableSlugSet.has(vslug);
      const target = linkable ? `[[tables/${vslug}]]` : `\`loom_views.${v.name}\``;
      parts.push(`- ${target}${v.description ? ` — ${v.description}` : ""}`);
    }
    parts.push("");
  }

  // ── Skills that mention this table ───────────────────────────────────
  const relatedSkills = skills.filter((s) =>
    new RegExp(`\\b${escapeReg(table.table_name)}\\b`, "i").test(s.description + " " + s.triggers.join(" ")),
  );
  if (relatedSkills.length > 0) {
    parts.push("## Used by skills");
    parts.push("");
    for (const s of relatedSkills) parts.push(`- **${s.name}** — ${s.description}`);
    parts.push("");
  }

  return parts.join("\n");
}

function renderIndexPage(tables: TableInfo[], views: ViewInfo[], skills: SkillInfo[]): string {
  const parts: string[] = [];
  parts.push(`# Tables overview`);
  parts.push("");
  parts.push(
    `Loom is indexing **${tables.filter((t) => !t.is_view).length} tables** and ` +
      `**${tables.filter((t) => t.is_view).length} saved views**. ` +
      `Click any link to drill into the page Loom generated for that object.`,
  );
  parts.push("");

  // Group by FK cluster (connected components of the FK graph).
  const clusters = clusterByFK(tables);

  for (const [i, group] of clusters.entries()) {
    parts.push(`## Cluster ${i + 1}: ${describeCluster(group)}`);
    parts.push("");
    for (const t of group) {
      const slug = tableSlug(t.schema_name, t.table_name);
      const meta = t.is_view ? "view" : `${t.row_count ?? 0} rows`;
      parts.push(`- [[tables/${slug}]] — ${meta}`);
    }
    parts.push("");
  }

  if (views.length > 0) {
    parts.push("## Saved views");
    parts.push("");
    for (const v of views) {
      const vslug = tableSlug("loom_views", v.name);
      parts.push(`- [[tables/${vslug}]]${v.description ? ` — ${v.description}` : ""}`);
    }
    parts.push("");
  }

  if (skills.length > 0) {
    parts.push("## Skills tying things together");
    parts.push("");
    for (const s of skills) parts.push(`- **${s.name}** — ${s.description}`);
    parts.push("");
  }

  return parts.join("\n");
}

// ─── Data loaders ──────────────────────────────────────────────────────────

async function loadTables(): Promise<TableInfo[]> {
  const r = await catalogPool.query<{
    table_id: number;
    schema_name: string;
    table_name: string;
    status: string;
    row_count: string | null;
    is_view: boolean;
    doc_md: string | null;
    fk_targets: string[] | null;
  }>(
    `SELECT t.id AS table_id,
            t.schema_name,
            t.table_name,
            t.status,
            t.row_count::text AS row_count,
            (t.schema_name = 'loom_views') AS is_view,
            d.markdown AS doc_md,
            COALESCE(
              (SELECT array_agg(c.fk_target ORDER BY c.ordinal)
                 FROM columns c
                WHERE c.table_id = t.id AND c.fk_target IS NOT NULL),
              ARRAY[]::text[]
            ) AS fk_targets
       FROM tables t
       LEFT JOIN docs d ON d.table_id = t.id
      ORDER BY t.schema_name, t.table_name`,
  );
  return r.rows.map((row) => ({
    table_id: row.table_id,
    schema_name: row.schema_name,
    table_name: row.table_name,
    status: row.status,
    row_count: row.row_count !== null ? Number(row.row_count) : null,
    is_view: row.is_view,
    doc_md: row.doc_md,
    fk_targets: row.fk_targets ?? [],
  }));
}

async function loadViews(): Promise<ViewInfo[]> {
  const r = await catalogPool.query<{ name: string; description: string | null; sql: string }>(
    `SELECT name, description, sql FROM proposals WHERE kind = 'view' AND status = 'applied'`,
  );
  return r.rows;
}

async function loadSkills(): Promise<SkillInfo[]> {
  const r = await catalogPool.query<{ slug: string; name: string; triggers: unknown; description: string }>(
    `SELECT slug, name, triggers, description FROM skills WHERE enabled = TRUE`,
  );
  return r.rows.map((row) => ({
    slug: row.slug,
    name: row.name,
    triggers: Array.isArray(row.triggers) ? (row.triggers as string[]) : [],
    description: row.description,
  }));
}

// ─── Helpers ───────────────────────────────────────────────────────────────

function tableSlug(schema: string, name: string): string {
  return `${schema}.${name}`.toLowerCase().replace(/[^a-z0-9._-]/g, "_");
}

function escapeReg(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/** Connected-components clustering over FK edges. */
function clusterByFK(tables: TableInfo[]): TableInfo[][] {
  const idx = new Map<string, number>();
  tables.forEach((t, i) => idx.set(`${t.schema_name}.${t.table_name}`, i));

  // Disjoint-set
  const parent = tables.map((_, i) => i);
  const find = (x: number): number => (parent[x] === x ? x : (parent[x] = find(parent[x]!)));
  const union = (a: number, b: number) => {
    const ra = find(a);
    const rb = find(b);
    if (ra !== rb) parent[ra] = rb;
  };

  tables.forEach((t, i) => {
    for (const fk of t.fk_targets) {
      const m = fk.match(/^([^.]+)\.([^.]+)/);
      if (!m) continue;
      const targetKey = `${m[1]}.${m[2]}`;
      const j = idx.get(targetKey);
      if (j !== undefined) union(i, j);
    }
  });

  const buckets = new Map<number, TableInfo[]>();
  tables.forEach((t, i) => {
    const r = find(i);
    if (!buckets.has(r)) buckets.set(r, []);
    buckets.get(r)!.push(t);
  });
  // Sort: bigger clusters first.
  return [...buckets.values()].sort((a, b) => b.length - a.length);
}

function describeCluster(group: TableInfo[]): string {
  // Crude: pick the table with the most FK refs as the centerpiece, plus
  // count of members.
  const sorted = [...group].sort((a, b) => b.fk_targets.length - a.fk_targets.length);
  const lead = sorted[0]!.table_name;
  return `${lead} & ${group.length - 1} related table${group.length === 2 ? "" : "s"}`;
}

File: backend/app/workers/wiki/tables.py
"""Tables-wiki agent.

Reads the catalog's tables/views/skills and writes interlinked concept pages
into wiki_pages with kind='tables'. Per-table pages plus a clustered index
page; hash-gated end-to-end via ``upsert_wiki_page``.
"""
from __future__ import annotations

import logging
import re
from dataclasses import dataclass, field
from typing import Any

from psycopg.rows import dict_row

from ...catalog.wiki import upsert_wiki_page
from ...db import get_conn

log = logging.getLogger(__name__)

ACTOR = "wiki-agent:tables"


@dataclass(slots=True)
class TableInfo:
    table_id: int
    schema_name: str
    table_name: str
    status: str
    row_count: int | None
    is_view: bool
    doc_md: str | None
    fk_targets: list[str] = field(default_factory=list)


@dataclass(slots=True)
class ViewInfo:
    name: str
    description: str | None
    sql: str


@dataclass(slots=True)
class SkillInfo:
    slug: str
    name: str
    triggers: list[str]
    description: str


async def run_tables_wiki_agent() -> dict[str, int]:
    tables = await _load_tables()
    views = await _load_views()
    skills = await _load_skills()

    table_slug_set: set[str] = {_table_slug(t.schema_name, t.table_name) for t in tables}

    generated = 0

    # ── Per-table / per-view pages ───────────────────────────────────────────
    for t in tables:
        slug = _table_slug(t.schema_name, t.table_name)
        title = (
            f"{t.schema_name}.{t.table_name}  (view)"
            if t.is_view
            else f"{t.schema_name}.{t.table_name}"
        )
        body = _render_table_page(t, table_slug_set, views, skills)
        summary = (
            f"Saved view · {t.row_count or 0} rows"
            if t.is_view
            else f"Table · {t.row_count or 0} rows"
        )
        r = await upsert_wiki_page(
            ACTOR,
            kind="tables",
            slug=slug,
            title=title,
            summary=summary,
            body_md=body,
            source_ref={
                "table_id": t.table_id,
                "schema": t.schema_name,
                "name": t.table_name,
                "is_view": t.is_view,
            },
        )
        if r["action"] != "skipped":
            generated += 1

    # ── Index page ──────────────────────────────────────────────────────────
    index_body = _render_index_page(tables, views, skills)
    real_tables = sum(1 for t in tables if not t.is_view)
    saved_views = sum(1 for t in tables if t.is_view)
    ix = await upsert_wiki_page(
        ACTOR,
        kind="tables",
        slug="_index",
        title="Tables overview",
        summary=f"{real_tables} tables · {saved_views} saved views",
        body_md=index_body,
        source_ref={
            "tables": len(tables),
            "views": len(views),
            "skills": len(skills),
        },
    )
    if ix["action"] != "skipped":
        generated += 1

    return {"generated": generated}


# ── Render helpers ──────────────────────────────────────────────────────────

def _render_table_page(
    table: TableInfo,
    table_slug_set: set[str],
    views: list[ViewInfo],
    skills: list[SkillInfo],
) -> str:
    parts: list[str] = []
    parts.append(f"# {table.schema_name}.{table.table_name}")
    parts.append("")

    if table.doc_md:
        stripped = re.sub(r"^#\s+[^\n]+\n+", "", table.doc_md)
        parts.append(stripped)
        parts.append("")
    else:
        parts.append(
            "*The catalog hasn't produced a structural+semantic doc yet — "
            "re-tick the scheduler.*"
        )
        parts.append("")

    # ── See also (FK neighbours) ────────────────────────────────────────────
    fk_neighbours: list[str] = []
    for f in table.fk_targets:
        m = re.match(r"^([^.]+)\.([^.]+)", f)
        if not m:
            continue
        slug = _table_slug(m.group(1), m.group(2))
        if slug in table_slug_set:
            fk_neighbours.append(slug)

    if fk_neighbours:
        parts.append("## See also")
        parts.append("")
        for s in fk_neighbours:
            parts.append(f"- [[tables/{s}]]")
        parts.append("")

    # ── Used in views ───────────────────────────────────────────────────────
    name_re = re.compile(rf"\b{re.escape(table.table_name)}\b", re.IGNORECASE)
    used_in_views = [v for v in views if name_re.search(v.sql)]
    if used_in_views:
        parts.append("## Used in views")
        parts.append("")
        for v in used_in_views:
            vslug = _table_slug("loom_views", v.name)
            linkable = vslug in table_slug_set
            target = f"[[tables/{vslug}]]" if linkable else f"`loom_views.{v.name}`"
            tail = f" — {v.description}" if v.description else ""
            parts.append(f"- {target}{tail}")
        parts.append("")

    # ── Skills that mention this table ──────────────────────────────────────
    related_skills: list[SkillInfo] = []
    for sk in skills:
        haystack = sk.description + " " + " ".join(sk.triggers)
        if name_re.search(haystack):
            related_skills.append(sk)
    if related_skills:
        parts.append("## Used by skills")
        parts.append("")
        for sk in related_skills:
            parts.append(f"- **{sk.name}** — {sk.description}")
        parts.append("")

    return "\n".join(parts)


def _render_index_page(
    tables: list[TableInfo], views: list[ViewInfo], skills: list[SkillInfo]
) -> str:
    parts: list[str] = []
    real_tables = sum(1 for t in tables if not t.is_view)
    saved_views = sum(1 for t in tables if t.is_view)

    parts.append("# Tables overview")
    parts.append("")
    parts.append(
        f"Loom is indexing **{real_tables} tables** and "
        f"**{saved_views} saved views**. "
        "Click any link to drill into the page Loom generated for that object."
    )
    parts.append("")

    clusters = _cluster_by_fk(tables)
    for i, group in enumerate(clusters):
        parts.append(f"## Cluster {i + 1}: {_describe_cluster(group)}")
        parts.append("")
        for t in group:
            slug = _table_slug(t.schema_name, t.table_name)
            meta = "view" if t.is_view else f"{t.row_count or 0} rows"
            parts.append(f"- [[tables/{slug}]] — {meta}")
        parts.append("")

    if views:
        parts.append("## Saved views")
        parts.append("")
        for v in views:
            vslug = _table_slug("loom_views", v.name)
            tail = f" — {v.description}" if v.description else ""
            parts.append(f"- [[tables/{vslug}]]{tail}")
        parts.append("")

    if skills:
        parts.append("## Skills tying things together")
        parts.append("")
        for s in skills:
            parts.append(f"- **{s.name}** — {s.description}")
        parts.append("")

    return "\n".join(parts)


# ── Data loaders ────────────────────────────────────────────────────────────

async def _load_tables() -> list[TableInfo]:
    sql = """
        SELECT t.id AS table_id,
               t.schema_name,
               t.table_name,
               t.status,
               t.row_count::text AS row_count,
               (t.schema_name = 'loom_views') AS is_view,
               d.markdown AS doc_md,
               COALESCE(
                 (SELECT array_agg(c.fk_target ORDER BY c.ordinal)
                    FROM columns c
                   WHERE c.table_id = t.id AND c.fk_target IS NOT NULL),
                 ARRAY[]::text[]
               ) AS fk_targets
          FROM tables t
          LEFT JOIN docs d ON d.table_id = t.id
         ORDER BY t.schema_name, t.table_name
    """
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(sql)
            rows = await cur.fetchall()
    out: list[TableInfo] = []
    for row in rows:
        rc_raw = row["row_count"]
        rc: int | None = int(rc_raw) if rc_raw is not None else None
        out.append(
            TableInfo(
                table_id=int(row["table_id"]),
                schema_name=row["schema_name"],
                table_name=row["table_name"],
                status=row["status"],
                row_count=rc,
                is_view=bool(row["is_view"]),
                doc_md=row["doc_md"],
                fk_targets=list(row["fk_targets"] or []),
            )
        )
    return out


async def _load_views() -> list[ViewInfo]:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                "SELECT name, description, sql FROM proposals "
                "WHERE kind = 'view' AND status = 'applied'"
            )
            rows = await cur.fetchall()
    return [
        ViewInfo(name=r["name"], description=r["description"], sql=r["sql"]) for r in rows
    ]


async def _load_skills() -> list[SkillInfo]:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                "SELECT slug, name, triggers, description FROM skills WHERE enabled = TRUE"
            )
            rows = await cur.fetchall()
    out: list[SkillInfo] = []
    for r in rows:
        triggers_raw: Any = r["triggers"]
        triggers: list[str] = list(triggers_raw) if isinstance(triggers_raw, list) else []
        out.append(
            SkillInfo(
                slug=r["slug"],
                name=r["name"],
                triggers=triggers,
                description=r["description"],
            )
        )
    return out


# ── Helpers ────────────────────────────────────────────────────────────────

_SLUG_SAFE = re.compile(r"[^a-z0-9._-]")


def _table_slug(schema: str, name: str) -> str:
    return _SLUG_SAFE.sub("_", f"{schema}.{name}".lower())


def _cluster_by_fk(tables: list[TableInfo]) -> list[list[TableInfo]]:
    """Connected-components clustering over the FK graph."""
    idx: dict[str, int] = {}
    for i, t in enumerate(tables):
        idx[f"{t.schema_name}.{t.table_name}"] = i

    parent = list(range(len(tables)))

    def find(x: int) -> int:
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(a: int, b: int) -> None:
        ra, rb = find(a), find(b)
        if ra != rb:
            parent[ra] = rb

    for i, t in enumerate(tables):
        for fk in t.fk_targets:
            m = re.match(r"^([^.]+)\.([^.]+)", fk)
            if not m:
                continue
            target_key = f"{m.group(1)}.{m.group(2)}"
            j = idx.get(target_key)
            if j is not None:
                union(i, j)

    buckets: dict[int, list[TableInfo]] = {}
    for i, t in enumerate(tables):
        r = find(i)
        buckets.setdefault(r, []).append(t)

    return sorted(buckets.values(), key=lambda g: -len(g))


def _describe_cluster(group: list[TableInfo]) -> str:
    sorted_g = sorted(group, key=lambda t: -len(t.fk_targets))
    lead = sorted_g[0].table_name
    n = len(group) - 1
    suffix = "" if len(group) == 2 else "s"
    return f"{lead} & {n} related table{suffix}"
